pkg/engine/systemroleassignments.go

```diff
@@ -88,13 +92,35 @@ func createKubernetesMasterRoleAssignmentForAgentPools(masterProfile *api.Master
 		dependenciesToMasterVms[masterIdx] = fmt.Sprintf("[concat(variables('masterVMNamePrefix'), %d)]", masterIdx)
 	}
 
-	var roleAssignmentsForAllAgentPools = make([]DeploymentWithResourceGroupARM, len(agentPoolProfiles))
+	roleAssignmentsForAllAgentPools := []DeploymentWithResourceGroupARM{}
 
 	// The following is based on:
 	// * https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/role-based-access-control/role-assignments-template.md#create-a-role-assignment-at-a-resource-scope
 	// * https://github.com/Azure/azure-quickstart-templates/blob/master/201-rbac-builtinrole-multipleVMs/azuredeploy.json#L79
-	for agentPoolIdx, agentPool := range agentPoolProfiles {
+
+	// We're gonna keep track of distinct VNETs in use across all our node pools;
+	//   we only want to define master VM --> VNET role assignments once per VNET.
+	// If our cluster configuration includes more than one pool sharing a common VNET,
+	//   we define the master VM --> VNET role assignments (one per master VM) just once for those pools
+	var vnetInCluster = struct{}{}
+	vnets := make(map[string]struct{})
+	for _, agentPool := range agentPoolProfiles {
 		var roleAssignments = make([]interface{}, masterProfile.Count)
+		subnetElements := strings.Split(agentPool.VnetSubnetID, "/")
+		// We expect a very specific string format for the VnetSubnetID property;
```

Comment:
We could instead use a capturing regex here (and the other places we're splitting into elements by `/`). Could be safer and more self-documenting. 

Happy to help with that here or in a follow-on PR if you agree.

Reply:
What's that regex look like? I think it'll definitely be more concise, less convinced about more readable/self-documenting, but maybe that's a lack of imagination on my part.

Reply:
Something like:

```regex
(?P<BASE>\/subscriptions\/.*\/resourceGroups\/.*\/providers\/Microsoft\.Network\/virtualNetworks\/.*\/subnets)\/.*
```

But wow I forgot about forward slashes needing escaping...ick. And go requires a few lines of code to extract a named capturing expression. So maybe never mind.

Reply:
Right, I actually do think that's maybe slightly better, but let's merge and see how annoyed we get with maintaining it in this more brute force way.